### ISSUE TITLE
init: Assertion 'hashmap_isempty(m->jobs)' failed at src/core/manager.c:1474, function manager_clear_jobs_and_units(). Aborting.

### ISSUE THREAD

**Issue by frebib** [systemd/systemd#24577]
### systemd version the issue has been seen with

251.3

### Used distribution

Debian Bullseye

### Linux kernel version used

5.15.46

### CPU architectures issue was seen on

x86_64

### Expected behaviour you didn't see

No crash

### Unexpected behaviour you saw

```
Broadcast message from systemd-journald@133m11 (Tue 2022-09-06 12:18:06 UTC):

systemd[1]: Caught <ABRT> from our own process.


Broadcast message from systemd-journald@133m11 (Tue 2022-09-06 12:18:06 UTC):

systemd[1]: Caught <ABRT>, dumped core as pid 3960.


Broadcast message from systemd-journald@133m11 (Tue 2022-09-06 12:18:06 UTC):

systemd[1]: Freezing execution.

Sep 06 12:18:06 133m11 systemd[1]: Reloading.
Sep 06 12:18:06 133m11 systemd[1]: Assertion 'hashmap_isempty(m->jobs)' failed at src/core/manager.c:1474, function manager_clear_jobs_and_units(). Aborting.
Sep 06 12:18:06 133m11 systemd[1]: Caught <ABRT> from our own process.
Sep 06 12:18:06 133m11 systemd[1]: Caught <ABRT>, dumped core as pid 3960.
Sep 06 12:18:06 133m11 systemd[1]: Freezing execution.
```

### Steps to reproduce the problem

I seemed to manage to trigger this twice during early-boot (over serial console) by calling `systemctl stop` on two service units and one path unit, right as some initialisation routine calls `systemctl daemon-reload`.
I'm not sure how to put the system into debug mode from boot. I can tweak the kernel cmdline if necessary, but this happens so early that it's hard to do anything before the window passes.

**Comment by poettering** [comment]
Please provide a full backtrace of the generated coredump. i.e. open the coredump in gdb and then do "bt full", with debug symbols installed.

**Comment by luckyfunky** [comment]
We currently facing the same issue. However, the bug occurs rarely: sometimes 1 in 100 reboots, sometimes 1 in 500 reboots. Since we usually build our firmware without coredump support I cannot provide any coredump.
When we enable coredump support the bug seems to disappear.

**Comment by keszybz** [comment]
I'll unset the milestone then. There might be a bug, but with the currently available information we are never going to figure this out. Somebody who can reproduce the issue will need to help.

**Comment by evverx** [comment]
> We currently facing the same issue. However, the bug occurs rarely: sometimes 1 in 100 reboots, sometimes 1 in 500 reboots

@luckyfunky I wonder if `systemd` was `>= 249` at the time?

**Comment by luckyfunky** [comment]
> > We currently facing the same issue. However, the bug occurs rarely: sometimes 1 in 100 reboots, sometimes 1 in 500 reboots
> 
> @luckyfunky I wonder if `systemd` was `>= 249` at the time?

Yes, the version is  250.5+

**Comment by evverx** [comment]
> Yes, the version is 250.5+

https://github.com/systemd/systemd/pull/25387 can probably address this issue then because it seems to touch the part of the code that can potentially lead to the assertion. I'm not 100% sure though because I haven't been able to trigger this particular assertion. https://github.com/systemd/systemd/issues/24990 and its variations seem to always lead to plain segfaults (without ASan) and they are somewhat always reproducible.

Anyway I was wondering about whether it was systemd >= 249 because it it wasn't it would most likely be a separate bug spanning more systemd versions.

**Comment by Paxy** [comment]
Appears that with v253 the bug still freeze system on reboot:
```
root@MVG-48210B357CF1:~# journalctl -e -n 10 -b -1

Apr 19 09:26:35 MVG-48210B357CF1 audit: BPF prog-id=0 op=UNLOAD
Apr 19 09:26:35 MVG-48210B357CF1 systemd[1]: Assertion 'hashmap_isempty(m->jobs)' failed at src/core/manager.c:1483, function manager_clear_jobs_and_units(). Aborting.
Apr 19 09:26:35 MVG-48210B357CF1 systemd[1]: Caught <ABRT> from our own process.
Apr 19 09:26:35 MVG-48210B357CF1 audit[4356]: ANOM_ABEND auid=4294967295 uid=0 gid=0 ses=4294967295 subj==unconfined pid=4356 comm="systemd" exe="/lib/systemd/systemd" sig=6 res=1
Apr 19 09:26:35 MVG-48210B357CF1 systemd-coredump[4357]: Due to PID 1 having crashed coredump collection will now be turned off.
Apr 19 09:26:35 MVG-48210B357CF1 systemd-coredump[4357]: 8;;man:core(5)^G[🡕]8;;^G Process 4356 (systemd) of user 0 dumped core.
                                                         
         Stack trace of thread 4356:
         #0  0x00007f890fe7a2e9 syscall (libc.so.6 + 0xf62e9)
         #1  0x00007f89101c8964 missing_rt_tgsigqueueinfo (libsystemd-shared-253.so + 0x24e964)
         #2  0x00007f89101c9463 propagate_signal (libsystemd-shared-253.so + 0x24f463)
         #3  0x000055856fbeccba crash (systemd + 0x15cba)
         #4  0x00007f890fdbcd60 n/a (libc.so.6 + 0x38d60)
         #5  0x00007f890fdbcce1 raise (libc.so.6 + 0x38ce1)
         #6  0x00007f890fda6537 abort (libc.so.6 + 0x22537)
         #7  0x00007f89101a558d log_assert_failed (libsystemd-shared-253.so + 0x22b58d)
         #8  0x00007f89104f735c manager_clear_jobs_and_units (libsystemd-core-253.so + 0x12935c)
         #9  0x00007f89104f73f9 manager_free (libsystemd-core-253.so + 0x1293f9)
         #10 0x000055856fbec42c main (systemd + 0x1542c)
         #11 0x00007f890fda7d0a __libc_start_main (libc.so.6 + 0x23d0a)
         #12 0x000055856fbddfca _start (systemd + 0x6fca)
         ELF object binary architecture: AMD x86-64
Apr 19 09:26:35 MVG-48210B357CF1 systemd[1]: Caught <ABRT>, dumped core as pid 4356.
Apr 19 09:26:35 MVG-48210B357CF1 systemd[1]: Freezing execution.
Apr 19 09:27:21 MVG-48210B357CF1 systemd-journald[696]: Failed to send WATCHDOG=1 notification message: Connection refused
root@MVG-48210B357CF1:~# systemd --version
systemd 253 (253)
+PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP -LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE -TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
```

**Comment by yuwata** [comment]
@Paxy Do you have any reproducer for the issue?

**Comment by Paxy** [comment]
I can't give reproduce steps as it is a pretty old system upgraded from old distros to the bullseye. But I can provide additional output.

**Comment by mrc0mmand** [comment]
> I can't give reproduce steps as it is a pretty old system upgraded from old distros to the bullseye. But I can provide additional output.

Do you have any SysV units on that machine? (I'm just curious if it's not another case of https://github.com/systemd/systemd/issues/24990)

**Comment by Paxy** [comment]
> Do you have any SysV units on that machine? (I'm just curious if it's not another case of #24990)

Not sure what do you mean, but when I search for sysv files, I can find:
```
/usr/bin/faked-sysv
/usr/bin/fakeroot-sysv
/usr/lib/x86_64-linux-gnu/libfakeroot/libfakeroot-sysv.so
/usr/sbin/sysv-rc-conf
```

**Comment by yuwata** [comment]
SysV files here mean old init scripts under e.g. /etc/init.d or rcX.d, depends on distro.

**Comment by yuwata** [comment]
Could you show the list of SysV init scripts and units under /run/systemd/system ?

**Comment by Paxy** [comment]
/run/systemd/system is empty, but /etc/init.d has many scripts.

**Comment by mrc0mmand** [comment]
> /run/systemd/system is empty, but /etc/init.d has many scripts.

Could you list them? Simple `ls -lR /etc/init.d` should be enough (assuming all the SysV scripts are from packages, not custom ones).

**Comment by Paxy** [comment]
```
root@MVG-48210B357CF1:~# ls -lR /etc/init.d
/etc/init.d:
total 384
lrwxrwxrwx 1 root root   36 Dec 16  2021 README -> /usr/share/doc/sysv-rc/init.d-README
-rwxr-xr-x 1 root root 2269 Nov 22  2018 acpid
-rwxr-xr-x 1 root root 5574 Nov  5  2019 alsa-utils
-rwxr-xr-x 1 root root 2055 May 19  2019 anacron
-rwxr-xr-x 1 root root 3740 Mar 30  2019 apparmor
-rwxr-xr-x 1 root root 1071 Jun 25  2011 atd
-rwxr-xr-x 1 root root 2401 Apr 13  2015 avahi-daemon
-rwxr-xr-x 1 root root 1175 Jan 24  2019 binfmt-support
-rwxr-xr-x 1 root root 2948 Aug 18  2014 bluetooth
-rwxr-xr-x 1 root root 1438 Dec 16  2021 bootlogs
-rwxr-xr-x 1 root root 1374 Dec 16  2021 bootmisc.sh
-rwxr-xr-x 1 root root 2178 Feb  3  2019 brightness
-rwxr-xr-x 1 root root 3565 Sep  9  2014 cgmanager
-rwxr-xr-x 1 root root 3354 Sep  9  2014 cgproxy
-rwxr-xr-x 1 root root 3815 Dec 16  2021 checkfs.sh
-rwxr-xr-x 1 root root 1072 Apr  6  2015 checkroot-bootclean.sh
-rwxr-xr-x 1 root root 9661 Dec 16  2021 checkroot.sh
-rwxr-xr-x 1 root root 1232 Aug 15  2019 console-setup.sh
-rwxr-xr-x 1 root root 3059 Jun 23  2019 cron
-rwxr-xr-x 1 root root  937 Aug 26  2019 cryptdisks
-rwxr-xr-x 1 root root  896 Aug 26  2019 cryptdisks-early
-rwxr-xr-x 1 root root 2804 Aug 21  2019 cups
-rwxr-xr-x 1 root root 1961 Jul  3  2015 cups-browsed
-rwxr-xr-x 1 root root 3152 Oct  5  2022 dbus
-rwxr-xr-x 1 root root 7167 May 23  2021 exim4
-rwxr-xr-x 1 root root 3029 Dec 18  2020 gdm3
-rwxr-xr-x 1 root root 3892 May 14  2014 gdm3.dpkg-old
-rwxr-xr-x 1 root root 1399 Feb  1  2019 gdomap
-rwxr-xr-x 1 root root 1336 Dec 16  2021 halt
-rwxr-xr-x 1 root root 3060 Jan 21  2018 hddtemp
-rwxr-xr-x 1 root root  927 Feb  3  2019 hostname.sh
-rwxr-xr-x 1 root root 1748 Jan 20  2022 hwclock.sh
-rwxr-xr-x 1 root root 2444 Feb  4  2019 irqbalance
-rwxr-xr-x 1 root root 1479 Oct 10  2016 keyboard-setup.sh
-rwxr-xr-x 1 root root 1300 Dec 16  2021 killprocs
-rwxr-xr-x 1 root root 2044 Feb  9  2019 kmod
-rwxr-xr-x 1 root root 2610 Feb 22  2019 lightdm
-rwxr-xr-x 1 root root  883 May 17  2016 lm-sensors
-rwxr-xr-x 1 root root 2661 May  2  2016 logmein-hamachi
-rwxr-xr-x 1 root root 2528 Feb 17  2012 minissdpd
-rwxr-xr-x 1 root root  821 Dec 16  2021 mount-configfs
-rwxr-xr-x 1 root root  677 Apr  6  2015 mountall-bootclean.sh
-rwxr-xr-x 1 root root 2983 Dec 16  2021 mountall.sh
-rwxr-xr-x 1 root root 1871 Dec 16  2021 mountdevsubfs.sh
-rwxr-xr-x 1 root root 1588 Dec 16  2021 mountkernfs.sh
-rwxr-xr-x 1 root root  685 Apr  6  2015 mountnfs-bootclean.sh
-rwxr-xr-x 1 root root 2433 Dec 16  2021 mountnfs.sh
-rwxr-xr-x 1 root root 4486 Sep 21  2020 networking
-rwxr-xr-x 1 root root 5658 Aug 13  2014 nfs-common
-rwxr-xr-x 1 root root 4579 Aug 13  2019 nginx
-rwxr-xr-x 1 root root 1494 Mar 21  2019 ntp
-rwxr-xr-x 1 root root 9138 Apr 28  2021 openvpn
-rwxr-xr-x 1 root root 3720 Oct 12  2018 pcscd
-rwxr-xr-x 1 root root 4485 Jan 25  2017 php5-fpm
-rwxr-xr-x 1 root root 1386 Mar  2  2021 plymouth
-rwxr-xr-x 1 root root  760 Mar  2  2021 plymouth-log
-rwxr-xr-x 1 root root  924 May 31  2018 procps
-rwxr-xr-x 1 root root  469 Feb 26  2021 pulseaudio-enable-autospawn
lrwxrwxrwx 1 root root   12 Dec 16  2021 rc -> /lib/init/rc
-rwxr-xr-x 1 root root  819 Dec 16  2021 rc.local
lrwxrwxrwx 1 root root   13 Dec 16  2021 rcS -> /lib/init/rcS
-rwxr-xr-x 1 root root  787 Dec 16  2021 reboot
-rwxr-xr-x 1 root root 3990 Jan  6  2021 resolvconf
-rwxr-xr-x 1 root root 1037 Oct 29  2018 rmnologin
-rwxr-xr-x 1 root root 2507 Jul 13  2020 rpcbind
-rwxr-xr-x 1 root root 4417 Mar 15  2019 rsync
-rwxr-xr-x 1 root root 2864 Feb 26  2019 rsyslog
-rwxr-xr-x 1 root root 2224 Apr 15  2018 saned
-rwxr-xr-x 1 root root 1222 May 21  2017 screen-cleanup
-rwxr-xr-x 1 root root 3207 Dec 16  2021 sendsigs
-rwxr-xr-x 1 root root  592 Oct 29  2018 single
-rwxr-xr-x 1 root root 3088 Oct 10  2019 smartmontools
-rwxr-xr-x 1 root root 2035 Nov 30 21:10 speech-dispatcher
-rwxr-xr-x 1 root root 4056 Jul  2  2022 ssh
-rwxr-xr-x 1 root root 1030 Jan 12  2019 sudo
-rwxr-xr-x 1 root root 1033 Oct 10  2019 tpm2-abrmd
-rwxr-xr-x 1 root root 6871 Aug  7  2022 udev
-rwxr-xr-x 1 root root 2083 Aug 15  2017 ufw
-rwxr-xr-x 1 root root 2781 Dec 16  2021 umountfs
-rwxr-xr-x 1 root root 2260 Dec 16  2021 umountnfs.sh
-rwxr-xr-x 1 root root 1234 Oct 26  2018 umountroot
-rwxr-xr-x 1 root root 1391 Jun  8  2019 unattended-upgrades
-rwxr-xr-x 1 root root 3111 Oct 15  2012 urandom
-rwxr-xr-x 1 root root 1588 Aug 21  2018 vnstat
-rwxr-xr-x 1 root root 4752 May 24  2016 wicd
-rwxr-xr-x 1 root root 2757 Nov 23  2016 x11-common

```

**Comment by yuwata** [comment]
Sorry, please also list units in /run/systemd/generator.

**Comment by Paxy** [comment]
```
root@MVG-48210B357CF1:~# ls -lR /run/systemd/generator
/run/systemd/generator:
total 16
-rw-r--r-- 1 root root 353 Apr 20 11:41 -.mount
-rw-r--r-- 1 root root 545 Apr 20 11:41 backup.mount
-rw-r--r-- 1 root root 387 Apr 20 11:41 boot-efi.mount
-rw-r--r-- 1 root root 543 Apr 20 11:41 boot.mount
drwxr-xr-x 2 root root 120 Apr 20 11:41 local-fs.target.requires
drwxr-xr-x 2 root root  80 Apr 20 11:41 local-fs.target.wants
drwxr-xr-x 2 root root  60 Apr 20 11:41 multi-user.target.wants
drwxr-xr-x 2 root root  60 Apr 20 11:41 openvpn.service.wants

/run/systemd/generator/local-fs.target.requires:
total 0
lrwxrwxrwx 1 root root 10 Apr 20 11:41 -.mount -> ../-.mount
lrwxrwxrwx 1 root root 15 Apr 20 11:41 backup.mount -> ../backup.mount
lrwxrwxrwx 1 root root 17 Apr 20 11:41 boot-efi.mount -> ../boot-efi.mount
lrwxrwxrwx 1 root root 13 Apr 20 11:41 boot.mount -> ../boot.mount

/run/systemd/generator/local-fs.target.wants:
total 0
lrwxrwxrwx 1 root root 45 Apr 20 11:41 systemd-fsck-root.service -> /lib/systemd/system/systemd-fsck-root.service
lrwxrwxrwx 1 root root 46 Apr 20 11:41 systemd-remount-fs.service -> /lib/systemd/system/systemd-remount-fs.service

/run/systemd/generator/multi-user.target.wants:
total 0
lrwxrwxrwx 1 root root 36 Apr 20 11:41 rc-local.service -> /lib/systemd/system/rc-local.service

/run/systemd/generator/openvpn.service.wants:
total 0
lrwxrwxrwx 1 root root 36 Apr 20 11:41 openvpn@client.service -> /lib/systemd/system/openvpn@.service

```

**Comment by yuwata** [comment]
Ah, sorry again. how about /run/systemd/generator.early and /run/systemd/generator.late ?

**Comment by Paxy** [comment]
```
oot@MVG-48210B357CF1:~# ls -lR /run/systemd/generator.early
ls: cannot access '/run/systemd/generator.early': No such file or directory
root@MVG-48210B357CF1:~# ls -lR /run/systemd/generator.late
/run/systemd/generator.late:
total 20
-rw-r--r-- 1 root root 769 Apr 20 11:48 exim4.service
-rw-r--r-- 1 root root 429 Apr 20 11:48 gdomap.service
drwxr-xr-x 2 root root 100 Apr 20 11:48 graphical.target.wants
-rw-r--r-- 1 root root 578 Apr 20 11:48 hddtemp.service
-rw-r--r-- 1 root root 479 Apr 20 11:48 logmein-hamachi.service
-rw-r--r-- 1 root root 608 Apr 20 11:48 minissdpd.service
drwxr-xr-x 2 root root 100 Apr 20 11:48 multi-user.target.wants

/run/systemd/generator.late/graphical.target.wants:
total 0
lrwxrwxrwx 1 root root 16 Apr 20 11:48 exim4.service -> ../exim4.service
lrwxrwxrwx 1 root root 18 Apr 20 11:48 hddtemp.service -> ../hddtemp.service
lrwxrwxrwx 1 root root 20 Apr 20 11:48 minissdpd.service -> ../minissdpd.service

/run/systemd/generator.late/multi-user.target.wants:
total 0
lrwxrwxrwx 1 root root 16 Apr 20 11:48 exim4.service -> ../exim4.service
lrwxrwxrwx 1 root root 18 Apr 20 11:48 hddtemp.service -> ../hddtemp.service
lrwxrwxrwx 1 root root 20 Apr 20 11:48 minissdpd.service -> ../minissdpd.service

```

**Comment by yuwata** [comment]
Thanks. Than, please provide the contents of .service units in /run/systemd/generator.late .

**Comment by Paxy** [comment]
[services.log](https://github.com/systemd/systemd/files/11284847/services.log)

**Comment by yuwata** [comment]
One spurious thing I found is that minissdpd provides both init file and native .service file... Maybe they conflict each other?

Is the issue reproducible when you disable and stop minissdpd.service ?

**Comment by Paxy** [comment]
Testing when the issue starts again. For some reason, the issue is not persistent. But when it starts freezing, it freezes almost every reboot.

**Comment by Paxy** [comment]
Yeah, looks like disabling this service triggers the bug:
```
root@MVG-48210B357CF1:~# systemctl disable minissdpd
minissdpd.service is not a native service, redirecting to systemd-sysv-install.
Executing: /lib/systemd/systemd-sysv-install disable minissdpd
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
	LANGUAGE = "en_US:en",
	LC_ALL = (unset),
	LC_TIME = "en_US.UTF-8",
	LC_MONETARY = "sr_RS",
	LC_ADDRESS = "sr_RS",
	LC_TELEPHONE = "sr_RS",
	LC_NAME = "sr_RS",
	LC_MEASUREMENT = "sr_RS",
	LC_IDENTIFICATION = "sr_RS",
	LC_NUMERIC = "sr_RS",
	LC_PAPER = "sr_RS",
	LANG = "en_US.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to a fallback locale ("en_US.UTF-8").
Failed to reload daemon: Connection reset by peer

Message from syslogd@MVG-48210B357CF1 at Apr 20 13:54:03 ...
 systemd[1]: Caught <ABRT>, dumped core as pid 3974.

Broadcast message from systemd-journald@MVG-48210B357CF1 (Thu 2023-04-20 13:54:03 CEST):

systemd[1]: Caught <ABRT>, dumped core as pid 3974.


Message from syslogd@MVG-48210B357CF1 at Apr 20 13:54:03 ...
 systemd[1]: Freezing execution.

Broadcast message from systemd-journald@MVG-48210B357CF1 (Thu 2023-04-20 13:54:03 CEST):

systemd[1]: Freezing execution.

perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
	LANGUAGE = "en_US:en",
	LC_ALL = (unset),
	LC_TIME = "en_US.UTF-8",
	LC_MONETARY = "sr_RS",
	LC_ADDRESS = "sr_RS",
	LC_TELEPHONE = "sr_RS",
	LC_NAME = "sr_RS",
	LC_MEASUREMENT = "sr_RS",
	LC_IDENTIFICATION = "sr_RS",
	LC_NUMERIC = "sr_RS",
	LC_PAPER = "sr_RS",
	LANG = "en_US.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to a fallback locale ("en_US.UTF-8").
insserv: warning: current start runlevel(s) (empty) of script `minissdpd' overrides LSB defaults (2 3 4 5).
insserv: warning: current stop runlevel(s) (0 1 2 3 4 5 6) of script `minissdpd' overrides LSB defaults (0 1 6).

Failed to reload daemon: Failed to activate service 'org.freedesktop.systemd1': timed out (service_start_timeout=25000ms)
```

**Comment by Paxy** [comment]
Removing all those services in /run/systemd/generator.late (except gdomap.service), still causes the freeze in reboot (v247, but same with v253):
```
Apr 20 19:38:54 MVG-48210B357CF1 systemd[1]: systemd-reboot.service: Succeeded.
Apr 20 19:38:54 MVG-48210B357CF1 systemd[1]: Finished Reboot.
Apr 20 19:38:54 MVG-48210B357CF1 systemd[1]: Reached target Reboot.
Apr 20 19:38:54 MVG-48210B357CF1 systemd[1]: Shutting down.
Apr 20 19:38:54 MVG-48210B357CF1 systemd[1]: Assertion 'hashmap_isempty(m->jobs)' failed at src/core/manager.c:1342, function manager_clear_jobs_and_units(). Aborting.
Apr 20 19:38:54 MVG-48210B357CF1 systemd[1]: Caught <ABRT>, dumped core as pid 4214.
Apr 20 19:38:54 MVG-48210B357CF1 systemd[1]: Freezing execution.
```

**Comment by Paxy** [comment]
Here is the complete journal log since the reboot command was issued until systemd crashed.
[journal.log](https://github.com/systemd/systemd/files/11289260/journal.log)

**Comment by yuwata** [comment]
> Yeah, looks like disabling this service triggers the bug:

Could you show the result of coredumpctl and/or journalctl of the ABRT on disabling the service?

**Comment by bryce-cahill** [comment]
I have a Yocto build that is also experiencing a crash with systemd v250. I have patched systemd with both #25387 and #26875; however, the bug can still be recreated. Bearing in mind that this is a private repository, I will have to redact specifics when providing any extra details so we will call the particular service in question "X". I can recreate the crash very consistently by doing the following: 

```
systemctl stop X
systemctl disable X
```

However, if you do this around the other way i.e.
```
systemctl disable X
systemctl stop X
```
it doesn't crash until you do another `systemctl disable X`

With this in mind I added some extra logging to the `manager_clear_jobs_and_units()` I can see that after the units are cleared there is one remaining job in the job hashmap  causing the assert. When testing further I can see that when I run `systemctl list-jobs` before and after the `systemctl stop X` I can see that the job in the list is replaced by two new job ID's. One of these jobs is the job that causes the assert.

If I try to cancel the job in question using `systemctl cancel <job-id>`, systemd crashes for a different reason:
```
[  192.558898] systemd[1]: Assertion '*pj == j' failed at src/core/job.c:144, function job_uninstall(). Aborting.
```

I hope this helps narrowing things down, let me know if there are any further details I can provide.

**Comment by yuwata** [comment]
> With this in mind I added some extra logging to the `manager_clear_jobs_and_units()` I can see that after the units are cleared there is one remaining job in the job hashmap causing the assert.

Could you provide the change and logs?

**Comment by yuwata** [comment]
Please also provide the service 'X'.

**Comment by mrc0mmand** [comment]
This is strange, I can't reproduce it even on Debian Sid/Experimental with the aforementioned SysV services no matter how hard I hammer systemd with enable/disable/start/stop requests.

> Yeah, looks like disabling this service triggers the bug:

Could you show `systemctl list-jobs` before doing the problematic `disable`?

**Comment by bryce-cahill** [comment]
I'll try to provide some more information on Monday however, I did have some other findings that may be helpful in the meantime. The system in question has a service that runs on boot, a number of other services wait for this service to complete before running, this includes service "x".

The crash happens if you disable service x while it is waiting for the first service. So at the time the disable is sent service x is loaded but inactive.

If I stop the first service so that service x has started as usual, the bug can no longer be recreated using the usual method of stop and disable.

**Comment by CarstenHansen616** [comment]
I work with @bryce-cahill on the affected Yocto project and wanted to share a few more findings.

We now suspect the issue to be caused by using a combination of a path unit and a service unit, having the service unit end up in the job list, and then stopping the service.

Again referring to our service as 'x', in our case we had

1. an x.path unit with
    PathExists=/some/existing/file
    Unit=x.service
2. x.service sitting in the job list in a 'waiting' state
3. a shutdown or other event stopping x.service (damage probably done at this point)
4. some other systemd operation happening

At step 2, `systemctl list-jobs` would show

```
JOB UNIT                                 TYPE  STATE
387 x.service                            start waiting
...
```

After step 3, `systemctl list-jobs` would show

```
JOB UNIT                                 TYPE  STATE
393 x.service                            start waiting
412 x.service                            start waiting
...
```

So after step 3 there would be _two_ entries in the job list referencing x.service, and both job IDs would be different from that in the previous listing.

The assertion failure could now be triggered e.g. by disabling a service (_any_ service, it would appear), shutting down or rebooting.

**Comment by yuwata** [comment]
Please provide details of x.service.

**Comment by yuwata** [comment]
Ah, finally I reproduced the issue.
```
$ head testsuite-63.units/test63-issue-24577*
==> testsuite-63.units/test63-issue-24577-dep.service <==
# SPDX-License-Identifier: LGPL-2.1-or-later
[Service]
Type=oneshot
ExecStart=bash -c 'sleep infinity'

==> testsuite-63.units/test63-issue-24577.path <==
# SPDX-License-Identifier: LGPL-2.1-or-later
[Path]
PathExists=/tmp/hoge

==> testsuite-63.units/test63-issue-24577.service <==
# SPDX-License-Identifier: LGPL-2.1-or-later
[Unit]
Requires=test63-issue-24577-dep.service
After=test63-issue-24577-dep.service

[Service]
Type=oneshot
ExecStart=bash -c 'sleep infinity'
```

```diff
diff --git a/test/units/testsuite-63.sh b/test/units/testsuite-63.sh
index 7ee7fc1513..5b7984317e 100755
--- a/test/units/testsuite-63.sh
+++ b/test/units/testsuite-63.sh
@@ -41,6 +41,22 @@ systemctl stop test63-glob.path test63-glob.service
 
 test "$(busctl --json=short get-property org.freedesktop.systemd1 /org/freedesktop/systemd1/unit/test63_2dglob_2eservice org.freedesktop.systemd1.Unit ActivationDetails)" = '{"type":"a(ss)","data":[]}'
 
+rm -f /tmp/hoge
+systemctl start test63-issue-24577.path
+systemctl status -n 0 test63-issue-24577.path
+systemctl status -n 0 test63-issue-24577.service || :
+systemctl list-jobs
+
+touch /tmp/hoge
+systemctl status -n 0 test63-issue-24577.path
+systemctl status -n 0 test63-issue-24577.service || :
+systemctl list-jobs
+
+systemctl stop test63-issue-24577.service
+systemctl status -n 0 test63-issue-24577.path
+systemctl status -n 0 test63-issue-24577.service || :
+systemctl list-jobs
+
 systemctl log-level info
 
 echo OK >/testok
```
Then, recently introduced assertion is triggered.
https://github.com/systemd/systemd/blob/aafb96d5146c70ac30f2b5c677cd22fd43abd62c/src/core/job.c#L262

**Comment by yuwata** [comment]
Memo, maybe caused by or related to 708961c7011b7578d3e6e10ca463d15edb62b5c1 ?

**Comment by yuwata** [comment]
Fix is waiting in #27451.
